### PR TITLE
NIH: rewrite getopt

### PR DESCRIPTION
**Pull request by bugaevc**
Fixes https://github.com/SerenityOS/serenity/issues/91

cc @alimpfard :wink:

### CODE REVIEW

**Comment by alimpfard** [comment]
I swear, if I see the silly globals that the old getopt had, I will...do nothing...and just nag about it

File: Libraries/LibC/getopt.cpp [hunk 1]
```diff
@@ -1,610 +1,355 @@
...
-#endif
-
-#define PRINT_ERROR	((opterr) && (*options != ':'))
-
-#define FLAG_PERMUTE	0x01	/* permute non-options to the end of argv */
-#define FLAG_ALLARGS	0x02	/* treat non-options as args to option "-1" */
-#define FLAG_LONGONLY	0x04	/* operate as getopt_long_only */
+int opterr = 1;
+int optopt = 0;
+int optind = 1;
```
Comment (alimpfard) [inline Libraries/LibC/getopt.cpp:35]: OH FFS

**Comment by alimpfard** [comment]
I like the idea, but can we have `OptionParser` not depend on the `optind` and co globals?
You could synchronise them at the end of `getopt` and `getopt_long`, and perhaps just keep a `g_option_parser` or something that tracks the global stupidity of getopt?

**Comment by bugaevc** [comment]
We could do that, but I thought that would be more boilerplate for little gain

**Comment by alimpfard** [comment]
can we at least document the behaviour then?
"to reset `getopt{_long,}`, please set `optind` to 0"
or maybe call `serenity_getopt_reset`?

**Comment by bugaevc** [comment]
> can we at least document the behaviour then?

Yes we can, and that reminds me, I forgot to write the man page, shame on me!

**Comment by bugaevc** [comment]
Now with `optreset` and `optind = 0` support, extensive man pages, colorful error messages, and misc stuff.

**Comment by bugaevc** [comment]
image

**Comment by alimpfard** [comment]
Very cool stuff!

**Pull request merged**